01-Login/src/app/auth/auth.service.ts

```diff
@@ -2,42 +2,127 @@ import { Injectable } from '@angular/core';
 import createAuth0Client from '@auth0/auth0-spa-js';
 import Auth0Client from '@auth0/auth0-spa-js/dist/typings/Auth0Client';
 import * as config from '../../../auth_config.json';
-import { BehaviorSubject } from 'rxjs';
+import { from, of, Observable, BehaviorSubject, combineLatest, throwError } from 'rxjs';
+import { tap, catchError, concatMap, shareReplay } from 'rxjs/operators';
+import { Router } from '@angular/router';
 
 @Injectable({
   providedIn: 'root'
 })
 export class AuthService {
-  isAuthenticated = new BehaviorSubject(false);
-  profile = new BehaviorSubject<any>(null);
+  // Create an observable of Auth0 instance of client
+  auth0Client$ = (from(
+    createAuth0Client({
+      domain: config.domain,
+      client_id: config.clientId,
+      redirect_uri: `${window.location.origin}/callback`
+    })
+  ) as Observable<Auth0Client>).pipe(
+    shareReplay(1), // Every subscription receives the same shared value
+    catchError(err => throwError(err))
+  );
+  // Define observables for SDK methods that return promises by default
+  // For each Auth0 SDK method, first ensure the client instance is ready
+  // concatMap: Using the client instance, call SDK method; SDK returns a promise
+  // from: Convert that resulting promise into an observable
+  isAuthenticated$ = this.auth0Client$.pipe(
+    concatMap((client: Auth0Client) => from(client.isAuthenticated()))
+  );
+  getUser$ = this.auth0Client$.pipe(
+    concatMap((client: Auth0Client) => from(client.getUser()))
+  );
+  handleRedirectCallback$ = this.auth0Client$.pipe(
+    concatMap((client: Auth0Client) => from(client.handleRedirectCallback()))
+  );
+  // Create subject and public observable of user profile data
+  private userProfileSubject$ = new BehaviorSubject<any>(null);
+  userProfile$ = this.userProfileSubject$.asObservable();
+  // Create a local property for login status
+  loggedIn: boolean = null;
 
-  private auth0Client: Auth0Client;
+  constructor(private router: Router) { }
 
-  config = config;
+  checkAuthOnInit() {
+    // This should only be called on app initialization
+    // Check if user already has an active session with Auth0
+    const checkAuth$ = this.isAuthenticated$.pipe(
```

Comment:
why do you need this during init? after `createAuth0Client`, you already know if you're authenticated or not, so I'm not sure why would you need to run this again.

Reply:
Wasn't aware of that feature of the SDK! I will update to remove it. Thanks!